Update engine with upstream changes 2024-05-31

aaronfranke opened the pull request:
This PR updates The Mirror's engine fork to be based on commit [`705b7a0b0bd535c95e4e8fb439f3d84b3fb4f427`](https://github.com/godotengine/godot/commit/705b7a0b0bd535c95e4e8fb439f3d84b3fb4f427) from 2024-05-31. This is about 2-3 months of upstream changes, and now this is Godot 4.3 beta instead of 4.3 dev. The commit that includes all of The Mirror's engine changes is [`5b3e2045d42a19a979c1357633ea0c350cbc3b60`](https://github.com/the-mirror-gdp/godot/commit/5b3e2045d42a19a979c1357633ea0c350cbc3b60), which is what the `project.godot` file has the hash set to. As usual, this is the head of [`themirror`](https://github.com/the-mirror-gdp/godot/tree/themirror).

I tested the basics, opening the app, logging in, joining a space, opening both a GDScript and a visual script in the editor, and executing script code. Overall things are pretty smooth sailing. As for changes to the app, one GraphEdit behavior change, one unknown problem with `fetch_callable`, and of course I reimported all files.

aaronfranke committed:
Update engine with upstream changes 2024-05-31

aaronfranke commented on mirror-godot-app/ui/common/components/endless_scroll_flow_container.gd:
```diff
@@ -69,6 +69,8 @@ func _calculate_max_items_on_screen(forced_size = Vector2.ZERO) -> int:
 
 
 func _fetch_data(cnt_items_to_fetch: int) -> Array:
+	if not fetch_callable.is_valid():
+		return []
 	assert(fetch_callable.is_valid())
```
@RevoluPowered Please investigate this before merge, the assert was being hit in the new engine builds.

<img width="840" alt="Screenshot 2024-06-02 at 3 20 16 AM" src="https://github.com/the-mirror-gdp/the-mirror/assets/1646875/e4fd74ca-824a-42bc-acc6-3c5ed5a7e2f3">

RevoluPowered replied:
Good to know checking this out now.

RevoluPowered replied:
it was a race condition caused by the code running before the user logged in. pushing up fix

RevoluPowered replied:
Fixed in my commit :) was simple

RevoluPowered committed:
fix endless scroll flow container race condition

caused by login, tested without the new if and it works fine but kept the if statement and added an error
added await for login
added minimum 1 element as it seemed wise

RevoluPowered commented on mirror-godot-app/ui/common/components/endless_scroll_flow_container.gd:
```diff
@@ -69,6 +69,11 @@ func _calculate_max_items_on_screen(forced_size = Vector2.ZERO) -> int:
+	if not fetch_callable.is_valid():
+		# this could be caused by the data not being initialised before this is called
+		# it happens generally if you're not logged in before populating this.
+		push_error("invalid fetch callable in endless scroll container")
```
I found another bug with this returning one element instead of the correct quantity i will investigate further

RevoluPowered replied:
Scroll container visibility and the initial fetch were fighting for their ordering, now it simply blocks calls until it has at least been configured once to rule out the possible race conditons

aaronfranke committed:
Update engine with Android Jolt fixes

RevoluPowered committed:
Fix broken scroll container quantity of tiles

RevoluPowered commented:
Tested / Fixed minor scroll container bug.

Weapon firing time is now from 5 seconds to 0.5 seconds on the first fire of the gun. Feels much more smooth on macbook requires wider testing and feels fine.
